DOCSUP-10300: Edited and translated to Russian (#25294)
ClickHouse/ClickHouse: docs/ru/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-sources.md

-   [Локальный файл](#dicts-external_dicts_dict_sources-local_file)

> -   [Исполняемый файл](#dicts-external_dicts_dict_sources-executable)
> -   [HTTP(s)](#dicts-external_dicts_dict_sources-http)
> -   [Исполняемый пул](#dicts-external_dicts_dict_sources-executable_pool)

Reviewer: ```suggestion
-   [Исполняемый файл](#dicts-external_dicts_dict_sources-executable)
-   [Исполняемый пул](#dicts-external_dicts_dict_sources-executable_pool)
-   [HTTP(s)](#dicts-external_dicts_dict_sources-http)
```

-   `command` — абсолютный путь к исполняемому файлу или имя файла (если каталог программы прописан в `PATH`).

> -   `format` — формат файла. Поддерживаются все форматы, описанные в разделе «[Форматы](../../../interfaces/formats.md#formats)».

Reviewer: ```suggestion
-   `format` — формат файла. Поддерживаются все форматы, описанные в разделе [Форматы](../../../interfaces/formats.md#formats).
```

-   `implicit_key` — исходный исполняемый файл может возвращать только значения, а соответствие запрошенным ключам определено неявно — порядком строк в результате. Значение по умолчанию: false.

> Этот источник словаря может быть настроен только с помощью XML-конфигурации. Создание словарей с исполняемым источником с помощью DDL отключено. Иначе пользователь базы данных сможет выполнить произвольный бинарный файл на узле ClickHouse.

Reviewer: ```suggestion
Этот источник словаря может быть настроен только с помощью XML-конфигурации. Создание словарей с исполняемым источником с помощью DDL запрещено. Иначе пользователь сможет выполнить произвольный бинарный файл на сервере ClickHouse.
```

## Исполняемый пул {#dicts-external_dicts_dict_sources-executable_pool}

> Исполняемый пул позволяет загружать данные из пула процессов. Этот источник не работает с макетами словарей, которые требуют загрузки всех данных источника. Исполняемый пул работает, если тип размещения словаря `cache`, `complex_key_cache`, `ssd_cache`, `complex_key_ssd_cache`, `direct`, `complex_key_direct`. Исполняемый пул генерирует пул процессов с указанной командой и оставляет их активными, пока те не выйдут. Программа считывает данные из потока STDIN, пока он доступен, и выводит результат в поток STDOUT, кроме того возможно ожидание следующего блока данных из STDIN. ClickHouse не закрывает STDIN после обработки блока данных, но посылает следующую часть данных, когда это требуется. Исполняемый скрипт должен быть готов к такому способу передачи данных — он должен опросить STDIN и сбросить данные в STDOUT заранее.

Reviewer: ```suggestion
Исполняемый пул позволяет загружать данные из пула процессов. Этот источник не работает со словарями, которые требуют загрузки всех данных из источника. Исполняемый пул работает словарями, которые размещаются [следующими способами](external-dicts-dict-layout.md#ways-to-store-dictionaries-in-memory): `cache`, `complex_key_cache`, `ssd_cache`, `complex_key_ssd_cache`, `direct`, `complex_key_direct`. 

Исполняемый пул генерирует пул процессов с помощью указанной команды и оставляет их активными, пока они не завершатся. Программа считывает данные из потока STDIN пока он доступен и выводит результат в поток STDOUT, а затем ожидает следующего блока данных из STDIN. ClickHouse не закрывает поток STDIN после обработки блока данных и отправляет в него следующую порцию данных, когда это требуется. Исполняемый скрипт должен быть готов к такому способу обработки данных — он должен заранее опрашивать STDIN и отправлять данные в STDOUT.
```

Пример настройки:

``` xml
<source>
    <executable_pool>
        <command><command>while read key; do printf "$key\tData for key $key\n"; done</command</command>
        <format>TabSeparated</format>
        <pool_size>10</pool_size>
        <max_command_execution_time>10<max_command_execution_time>
        <implicit_key>false</implicit_key>
    </executable_pool>
</source>
```

Поля настройки:

-   `command` — абсолютный путь к файлу или имя файла (если каталог программы записан в `PATH`).
-   `format` — формат файла. Поддерживаются все форматы, описанные в “[Форматы](../../../interfaces/formats.md#formats)”.
-   `pool_size` — размер пула. Если в поле `pool_size` указан 0, то размер пула не ограничен.

> -   `command_termination_timeout` — скрипт исполняемого пула, должен включать основной цикл чтения-записи. После уничтожения словаря канал закрывается. При этом исполняемый файл имеет `command_termination_timeout` секунд для завершения работы, прежде чем ClickHouse пошлет сигнал SIGTERM дочернему процессу. Указывается в секундах. Значение по умолчанию: 10. Необязательный параметр.

Reviewer: ```suggestion
-   `command_termination_timeout` — скрипт исполняемого пула должен включать основной цикл чтения-записи. После уничтожения словаря канал закрывается. При этом исполняемый файл имеет `command_termination_timeout` секунд для завершения работы, прежде чем ClickHouse пошлет сигнал SIGTERM дочернему процессу. Значение параметра указывается в секундах. Значение по умолчанию: 10. Необязательный параметр.
```
